### DISCUSSION TITLE
Write file failed when there is a lot of space.

### DISCUSSION THREAD

**Discussion by eliphatfs** [seaweedfs/seaweedfs#5818] [Q&A]
I have 1200x8=9600 volumes allocated. Due to whatever reason there were a lot of orphans after some time of usage. I just cleared the orphans with fsck and ran a vacuum after. There are 130T/260T available now, and most volumes are <8GB where the max is 30GB. However, when a new file is to be written, seaweedfs tries to allocate a new volume and refuses to write to these empty volumes.

1 master, 1 filer with 8 volume servers, linux, 3.65.

Jul 26 09:32:50 aries-login seaweedfs-master[3925619]: I0726 09:32:50.959089 filer_notify.go:104 metadata log write failed /topics/.system/log/2024-07-26/09-32.13b0a6e6: AssignVolume: rpc error: code = Unknown desc = no free volumes left for {"replication":{},"ttl":{"Count":0,"Unit":0}}

**Comment by eliphatfs** [comment]
[volume.txt](https://github.com/user-attachments/files/16389880/volume.txt)

**Comment by eliphatfs** [comment]
[PickForWrite](https://github.com/seaweedfs/seaweedfs/blob/master/weed/topology/volume_layout.go#L307)
It seems that on a truly random basis, when the system picked a volume between 90% and 100% full, it will intend to trigger a growth of volumes. In this case, the write will error out. However, I think as there are plenty of less crowded volumes, those should be written to instead of error.

**Comment by eliphatfs** [comment]
I think `ShouldGrowVolumes` need to be checked before returning error.

**Comment by eliphatfs** [comment]
Proposes a fix in #5819

**Reply by kmlebedev** [reply]
> Proposes a fix in #5819

this is different

**Reply by eliphatfs** [reply]
> > Proposes a fix in #5819
> 
> this is different

Sorry, what do you mean?

**Reply by kmlebedev** [reply]
> > > Proposes a fix in #5819
> > 
> > 
> > this is different
> 
> Sorry, what do you mean?

this is not fix

**Reply by kmlebedev** [reply]
Your catch this exception
```
			// if picked volume is almost full, trigger a volume-grow request
			if ms.Topo.AvailableSpaceFor(option) <= 0 {
				return nil, fmt.Errorf("no free volumes left for " + option.String())
			}
```

**Reply by eliphatfs** [reply]
I may see what you mean. But 
1) either we check before this return or check in the outer condition, the check will always be executed if other conditions in the outer branch (`shouldGrow && !vl.HasGrowRequest()`) meet.
2) even if we start a grow request, the master would discard the request in `ProcessGrowRequest`.
So it is better to place it in the outer branch.

**Comment by kmlebedev** [comment]
1. writables 0 
2. free 0
```
        Disk hdd(volume:1200/1200 active:1185 free:0 remote:0)

```

**Reply by eliphatfs** [reply]
there are lots of writables, free 0 yes we were stuffed with orphans by maybe another bug so all volumes were occupied. We removed the orphans, and many volumes have lots of space.

**Reply by eliphatfs** [reply]
In this case, there are a total of 130TB available, but still cannot write a 1GB file due to picking a crowded volume for one of the chunks and cannot grow more volumes. I don't think this is reasonable.

**Comment by kmlebedev** [comment]
What's reasonable? Describe the problem you want to solve at the moment.

**Reply by kmlebedev** [reply]
I think I got it, the problem is that the PickForWrite function found volumeLocationList, but because it required should Grow you hit `return nil, fmt.Errorf("no free volumes left for " + option.String())` although you could continue writing without growth

**Reply by eliphatfs** [reply]
Yes. And actually the master would discard the grow request because it thinks the growth is not necessary (active 1145 much larger than crowded ~10).

**Reply by eliphatfs** [reply]
So the request and the error should not be created in the first place.

**Comment by kmlebedev** [comment]
he error should be skipped, the behavior can be configured via `master.volume_growth.threshold`
